### DISCUSSION TITLE
Built-in report function should check user's settings and suggest to restore defaults? And to update lists?

### DISCUSSION THREAD

**Discussion by peace2000** [uBlockOrigin/uBlock-issues#2582] [Ideas]
I wonder what is the reason behind this great deal of "Unable to reproduce" reports at uAssets repo: https://github.com/uBlockOrigin/uAssets/issues?q=label%3A%22unable+to+reproduce%22+

I have a few ideas that could potentially help to reduce them.

- Suggest the user to update all filter lists before sending a report (in the reporting tool) and add a button to update lists, next to "Find similar reports" button

- Before a user sends an issue report through uBO's reporting function, maybe uBO should check that the user has defaults and ask to restore them before sending a report (but not to force)? Obviously, everyone isn't truthful in the form that they have tried with defaults

- Also uBO could additionally send a list of installed other browser addons along with uBO-related settings and filterlists in use to the issue report? This could help to figure out if an issue could be caused by other addons. And maybe in some cases ask to disable specific addons if that addon is known for causing trouble. Especially, if multiple adblockers are being used. (Though, volunteers can always just close the issue report if the issue can't be reproduced.)

**Comment by gorhill** [comment]
> uBO could additionally send a list of installed other browser addons

Not possible without adding a new [`management` permission](https://developer.chrome.com/docs/extensions/reference/management/) to the manifest, and I don't want to add such permission to uBO, people tend to react negatively when an extension asks for a new permission. It can be declared as an optional permission but I expect people to turn away when the browser ask to give a new permission to uBO and chose to not file a report.

For the two other ideas, I will give some more thought, I think a button to force an update of filter lists if they have not been recently updated would be useful.

**Reply by gorhill** [reply]
> I think a button to force an update of filter lists if they have not been recently updated would be useful.

https://github.com/gorhill/uBlock/commit/4a92f96206c2fba138c7323c3c4c9524ef71fa26

**Reply by stephenhawk8054** [reply]
Actually I'm afraid that people will just click the button and report anyway without checking the site again. That could give us false information about update time.

**Reply by gorhill** [reply]
I don't see the point of your comment. Are you saying we should give up on trying to add more guidance before reporting?

I think it will help -- otherwise I wouldn't have spent time on this. Once the lists update, there is a clear message to verify again (and I also entertained the idea of adding a reload button to force a reload of the tab and bring it forth). But for now we should not be pessimistic that this will help. It might be 100% foolproof, but it will definitely help.

At the very least, the troubleshooting information will tell us accurately now if they did skip that step.

**Reply by stephenhawk8054** [reply]
> I also entertained the idea of adding a reload button to force a reload of the tab and bring it forth

Yes, this is what I was also thinking about, since users can have clearer guidance of how to do it combining with the `Update now` button. I'm just not sure if there's any difficulty or issue if implementing it.

**Reply by krystian3w** [reply]
If they skip checking it refreshing page by reporter, we will still have FP that someone's updated filters continue to fail.

**Reply by peace2000** [reply]
> If they skip checking it refreshing page by reporter, we will still have FP that someone's updated filters continue to fail.

I think that implementing this change is better than not to implement. if a volunteer cannot reproduce the issue, it just will be closed as before. But I'm sure that this will help at least some degree.

**Reply by MasterKia** [reply]
> > uBO could additionally send a list of installed other browser addons

> Not possible without adding a new [management permission](https://developer.chrome.com/docs/extensions/reference/management/) to the manifest

@gorhill On Chromium you can (ab)use this method to detect the presence of other known content blockers:
https://github.com/arxenix/chrome-ublock-detection/blob/a4c93f705355354070b797f374b330fe548292ae/index.html#L36

Related issue:
https://github.com/uBlockOrigin/uBlock-issues/issues/1572

PoC:
https://arxenix.github.io/chrome-ublock-detection/

**Reply by gorhill** [reply]
I know about this, and assuming it still work from uBO's own page (I don't think so given extensions are not allowed to see each other's requests), I sure won't use this in uBO to try to detect other extensions, using such trick would just undermine trust in uBO.

**Reply by MasterKia** [reply]
It's no different than requesting the `Management` permission and accessing the list of _all_ installed extensions.

However we could also add a "Other extensions used" to the issue template where the user could fill it out.

Similar to AdGuard:
https://github.com/AdguardTeam/AdguardFilters/issues/153751

**Reply by Yuki2718** [reply]
At least permission is explicit to user.

**Reply by peace2000** [reply]
"Other extensions used" could imply that it's ok to send reports without disabling them.

**Reply by MasterKia** [reply]
> At least permission is explicit to user.

There's a trade-off to be made, and the trade-off in this case is wasting volunteers time by having other conflicting extensions.

**Reply by gorhill** [reply]
> It's no different than requesting the Management permission

It's doing something in a sneaky way which normally requires asking extensions API permission to users. Such behavior undermines trust, for good reasons. I would become suspicious of an extension doing this sort of thing to avoid asking my permission to do so.

However we could add a field for people to disclose other extensions used, and forcing them to write "none" if they don't have other extensions.

**Reply by MasterKia** [reply]
> and forcing them to write "none" if they don't have other extensions.

Cool, I didn't know GitHub supported that.

**Reply by gorhill** [reply]
https://github.com/uBlockOrigin/uAssets/commit/d066a4502113b09189155ebea84c2b309608ed6a

**Comment by krystian3w** [comment]
I, for one, wonder if detection of active cosmetics has a similar limit imposed as someone with 80 lists - so as not to scare people with GitHub's http error.

**Reply by ghost** [reply]
Good question. https://github.com/gorhill/uBlock/commit/6efd8eb84aaae43b05c3e9f2bffd13cf85f16018

I don't see any limit. The length of the URL may break new issue page again. @gorhill ?

**Reply by ghost** [reply]
For a moment I thought that this data can be packed, but it need to be plain text to be used in URL, so it gains only few percent when lz4 -> b64 :( (2.2 > 1.4 > 1.9 kB for ex.)

**Reply by gorhill** [reply]
I thought about the URL limit issue but I have no idea if long list of matched cosmetic filters will be a thing, so for now I opted for _"let's see how it goes"_.

**Reply by MasterKia** [reply]
> I have no idea if long list of matched cosmetic filters will be a thing

One case comes to mind, when some anti-blocker scripts try multiple bait cosmetics.

**Reply by krystian3w** [reply]
this was already checked and some of the classes/id as continue to exist in EasyList was undetected:

https://github.com/gorhill/uBlock/commit/6efd8eb84aaae43b05c3e9f2bffd13cf85f16018#commitcomment-107550105 https://github.com/gorhill/uBlock/commit/6efd8eb84aaae43b05c3e9f2bffd13cf85f16018#commitcomment-107552133

**Reply by gorhill** [reply]
I could change the indentation from 2-character to 1-character, this will reduce number of characters a bit.

**Reply by dimisa-RUAdList** [reply]
At the moment, after pressing the update filters button, this procedure is performed only for uBlock filters. Is it possible to make sure that all filters are updated, including those that were imported?

**Reply by gorhill** [reply]
I didn't include external filter lists because we do not support issues arising from them (we ask people to test without them before filing report), so there is no point updating them anyways.

**Reply by dimisa-RUAdList** [reply]
The fact is that users do not know in advance with which filter the problem occurs. And they somehow turn to the uBlock Origin bug tracker, no matter what happens to them.

Quite often it happens that the problem has already been solved a long time ago, it's just that the fixes were not pulled up during the scheduled auto-update.

Example: `https://github.com/uBlockOrigin/uAssets/issues/18131`, `https://github.com/uBlockOrigin/uAssets/issues/18139`

Cases like this take up the time of maintainers, and if there is a chance to avoid it, then why not take advantage of it?

**Reply by gorhill** [reply]
> The fact is that users do not know in advance with which filter the problem occurs

We ask in the prerequesite steps to see if the issue exists without external lists.

**Reply by dimisa-RUAdList** [reply]
Yes, I know that. But in most cases, users do not follow these recommendations.

**Reply by gorhill** [reply]
> procedure is performed only for uBlock filters

Wait, I thought you meant only for stock filters.

**Reply by gorhill** [reply]
It's not done just for uBO filters, it's done for all stock filters: <https://github.com/gorhill/uBlock/blob/1.49.3b17/src/js/messaging.js#L599-L603>. Imported filter list (aka "external") are left out.

**Reply by dimisa-RUAdList** [reply]
I just checked, and indeed, the update happened for all active standard filters. But the imported filters were not updated. Is it possible to enable the update for them too?

**Reply by gorhill** [reply]
We are going back to my argument that we ask people to test issues without imported (external) filter lists, as we do not want to spend time investigating issue with non stock lists.

**Reply by dimisa-RUAdList** [reply]
I understand this argument. But since the filters are updated anyway, why not update them all, as if it were done in the 3p-filters.html tab? In addition, by clicking the update button, the user gets the false confidence that he has updated all the filters, when in fact this is not the case.

**Reply by peace2000** [reply]
@dimisa-RUAdList I agree. Why deliberately only update stock lists as the same function could be used to update all the lists?

Many people are not truthful (this is checked frequently):

`I did not answer truthfully to ALL the above checkpoints.`

Updating external lists could solve issues and we would see less unnecessary issue reports. Lessening issue reports that are caused by outdated lists was the idea of this discussion after all.

**Comment by gorhill** [comment]
Looking at recent <https://github.com/uBlockOrigin/uAssets/issues> reports, I can see it does seem to help, but amazingly there are still reports which just ignore the request to update filter lists. Should we just disable the _Create new report_ button until the user did click _Update now_ and the lists have been all updated? We will still have to trust that the user verified the issue still exist after lists have updated.

**Reply by dimisa-RUAdList** [reply]
I think it's a good idea. Even if it works only partially, it will still lead to a decrease in cases of ignoring the recommendations.

But it is necessary that when you click the "Update now" button, ALL filters are updated.

**Reply by peace2000** [reply]
Yes for mandatory updating - for all lists.

**Reply by gorhill** [reply]
> ALL filters are updated

It is currently the case.

**Reply by ghost** [reply]
It's stupid that the "Create New Report" button is not available, and there is no explanation on how to enable it. The "Update Now" button hangs on Firefox ESR. I have the latest filters. 
What is a "good idea" to disable the button without explanation?

**Reply by krystian3w** [reply]
The problem is probably on your side (Internet, system or browser configuration (on ESR channel 115)), you can check which lists are outdated manually and the problem will disappear for about 1-5 hours with a frozen panel.

The only thing that can be improved is optional detection that someone has lists hosted on localhost - some error that the local server instead waiting user press refresh for tab after a few minutes (probably not many people use lists from localhost).

There may also be an error that even the title attribute with the words 'filter update required' still be useless with fresh lists (some error caused by RAM or CPU, difficult to address).

**Reply by ghost** [reply]
It doesn't enable the report button if the filters are up-to-date, and there is no explanation that you have to update filters again to enable the button.
One may decide that it just doesn't work and broken. Like me. And search this topic in google.
When I open the settings, the update button is not active, which means that filters are always up-to-date. But the report button is not.

**Reply by krystian3w** [reply]
Some default lists are considered old, some are updated more than 4 times a day, waiting for GitHub actions.

**Reply by u-RraaLL** [reply]
@migozz 

1. Screenshot the reporting widget showing your troubleshooting info.
2. Make sure you didn't disable default trusted sites.

> When I open the settings, the update button is not active, which means that filters are always up-to-date. But the report button is not.

Not true. Filter lists and the reporting widget have different definitions of "outdated". We don't want people to abuse updating filter lists (mainly the issue has arisen due to youtube anti adblock last year). At the same time we don't want to receive duplicate reports for things that have already been fixed, which is why the widget requires a manual update.

**Reply by garry-ut99** [reply]
Since some ideas are being implemented, what about this idea: https://github.com/uBlockOrigin/uBlock-issues/issues/1541#issuecomment-1432886015, I know there is "Find similiar reports" button, "but amazingly there are still reports which just ignore the request to find duplicate issues".

Or another alternative (or even additionally altogether with above idea) is to enhance already implemented idea: to disable "Create new report" button, not only until "Update filters" button is clicked and lists updated, but also until "Find similiar reports" button is clicked. "We will still have to trust that the user verified the duplicates don't exist"

Currently, I see there is a github-actions bot which closes duplicates https://github.com/uBlockOrigin/uAssets/issues/24472#issuecomment-2226893252, but it doesn't prevent or reduce amount of duplicates, only closes them and provides a link to the big issue.

**Reply by u-RraaLL** [reply]
Both ideas have been pitched before. The first one was rejected because (if I'm not mistaken) gorhill didn't want uBO to target any specific website within uBO's code (I'm not talking about scriptlets here). This could now be achieved using `uritransform` filters, but other team members were against that suggestion and were fine with the current bot actions.

For the second idea, I haven't received any feedback. Maybe because it was not a separate comment but a latter part of one that suggested renaming the YT megaissue back when they first started taking the detection globally.

**Reply by krystian3w** [reply]
In addition, tapping into other people's reports can be blocked by a private blacklist of first-time reporter e.g. 1+ year old issue.

Then someone may not feel the need to create an account for 7 days to discuss the problem and delete it account, so as not to irritate Microsoft.
